homeassistant/components/tag/__init__.py

```diff
@@ -133,7 +190,75 @@ async def async_scan_tag(
     )
 
     if tag_id in helper.data:
-        await helper.async_update_item(tag_id, {LAST_SCANNED: dt_util.utcnow()})
+        await helper.async_update_item(
+            tag_id, {LAST_SCANNED: dt_util.utcnow(), DEVICE_ID: device_id or ""}
+        )
     else:
-        await helper.async_create_item({TAG_ID: tag_id, LAST_SCANNED: dt_util.utcnow()})
+        await helper.async_create_item(
+            {TAG_ID: tag_id, LAST_SCANNED: dt_util.utcnow(), DEVICE_ID: device_id or ""}
+        )
     _LOGGER.debug("Tag: %s scanned by device: %s", tag_id, device_id)
+
+
+class TagEntity(Entity):
+    """Representation of a Tag entity."""
+
+    _unrecorded_attributes = frozenset({TAG_ID})
+    _attr_translation_key = DOMAIN
+    _attr_should_poll = False
+
+    # Implements it's own platform
+    _no_platform_reported = True
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        name: str,
+        tag_id: str,
+        last_scanned: str,
+        device_id: str | None,
+    ) -> None:
+        """Initialize the Tag event."""
+        self.entity_id = f"tag.{slugify(name)}"
```

Comment:
We should probably let the existing tooling pick the entity id

Reply:
This doesn't set the `entity_id`, this allows the integration to influence the `entity_id` picked by core.

Reply:
Right, but do we need this as isn't it nearly the same as what https://github.com/home-assistant/core/blob/85203aeb28fa8a31cfb905eec55e32dc505358be/homeassistant/helpers/entity_registry.py#L749 does already?